docs/DesignDocs/AttestationAPIsWithFlexibleClaims.md

```diff
@@ -0,0 +1,478 @@
+      size_t endorsements_buffer_size,
+      const oe_policy_t* policies,
+      size_t policies_size,
+      oe_claim_t** claims,
+      size_t* claims_length);
+  ```
+  The API does not introduce changes to the function prototype. The purpose of
+  the API is to pair with `oe_get_attestation_certificate_with_evidence_v3`;
+  i.e., only the API is able to verify the certificated generated via
+  `oe_get_attestation_certificate_with_evidence_v3` with either `runtime_custom_claims_buffer`
+  or `inittime_custom_claims_buffer` supplied.
+
+Design
+--------
+
+The new APIs follow the design outlined in
+[Open Enclave Init-time Configuration Interface](InitTimeConfigurationInterface.md)
+to support init-time custom claims.
+The core idea of the design is to avoid breaking changes to the attestation plugin
+framework. More specifically, the `oe_get_attestation_certificate_with_evidence_v3`
+API appends the `inittime_custom_claims_buffer` to the evidence buffer, which is the output
+of the `oe_get_evidence` API. After that, the API passes the new evidence buffer to the internal
+API that is used to generate the self-signed certificate with the evidence embedded.
+Similar to the run-time claims support, the API appends the `runtime_custom_claims_buffer` to the
+`public_key` and passes the new buffer to the `oe_get_evidence` API.
+
+```diff
+oe_result_t oe_get_attestation_certificate_with_evidence_v3(
+    const oe_uuid_t* format_id,
+    const unsigned char* subject_name,
+    const uint8_t* private_key,
+    size_t private_key_size,
+    const uint8_t* public_key,
+    size_t public_key_size,
++   const uint8_t* runtime_custom_claims_buffer,
++   size_t runtime_custom_claims_buffer_size,
++   const uint8_t* inittime_custom_claims_buffer,
++   size_t inittime_custom_claims_buffer_size,
+    const void* optional_parameters,
+    size_t optional_parameters_size,
+    uint8_t** output_certificate,
+    size_t* output_certificate_size)
+{
+    oe_result_t result = OE_FAILURE;
+    uint8_t* evidence_buffer = NULL;
+    size_t evidence_buffer_size = 0;
++   uint8_t* custom_claims_buffer = NULL;
++   size_t custom_claims_buffer_size = 0;
++   uint8_t* evidence_with_inittime_custom_claims = NULL;
++   size_t evidence_with_inittime_custom_claims_size = 0;
+
+    OE_TRACE_VERBOSE("Calling oe_get_attestation_certificate_with_evidence_v3");
+    OE_TRACE_VERBOSE(
+        "generate evidence with hash from public_key_size=%d public_key key "
+        "=\n[%s]\n",
+        public_key_size,
+        public_key);
+
++   custom_claims_buffer_size = public_key_size + runtime_custom_claims_buffer_size;
++   custom_claims_buffer = oe_malloc(custom_claims_buffer_size);
++   if (!custom_claims_buffer)
++       OE_RAISE(OE_OUT_OF_MEMORY);
+
++   memcpy(custom_claims_buffer, public_key, public_key_size);
++   memcpy(
++       custom_claims_buffer + public_key_size,
++       runtime_custom_claims_buffer,
++       runtime_custom_claims_buffer_size);
```

Comment:
The public key could have variable size. Once you copied it into the `custom_claims_buffer`, how do we extract it out?

Reply:
The design assumes the public_key_size is provided by the `oe_get_attestation_certificate_with_evidence_v3` API. The certificate generated by the API should include the public key with the given same size. The `oe_verify_attestation_certificate_with_evidence_v3` firstly gets the public key from the certificate and then extracts the runtime claim from the custom_claims_buffer.